### PR TITLE
ARCs, Relays, and Hubs

### PR DESCRIPTION
**Pull request by martindale**
Implement the complete user story for ARCs, Relays, and Hubs.

## Summary by CodeRabbit

- **New Features**
  - Added structured execution programs, opcode controls, contract capabilities, payment observation, withdrawals, Taproot policies, state attestations, and Bitcoin hallmarks.
  - Added contract synchronization, delivery tracking, encrypted group and onion-routed chat, blinded execution, Beacon safeguards, group governance, and safer sidechain registry patches.
  - Added interactive setup with encrypted wallets, locking, backups, restoration, and network-specific identity derivation.
  - Added wallet lock/unlock controls and expanded CLI setup options.

- **Bug Fixes & Security**
  - Improved authorization, signature validation, self-connection prevention, relay safety, message handling, and network compatibility checks.

- **Documentation**
  - Updated protocol, API, installation, security, identity, and release documentation. npm 12+ is now required.

---

> [!NOTE]
> **High Risk**
> Changes contract publish authorization, P2P document handling, and new Bitcoin Taproot spend paths—security-sensitive areas that affect funds and mesh trust.
> 
> **Overview**
> Adds **contract Taproot** tooling (`contractTaproot`, tier `when` predicates) plus **`Contract` / `Federation` `toAddress` / `toTaprootContract`** so spend policies compile to deterministic P2TR ladders (failover, decay/migrate, PSBT prep). Introduces **contract-scoped capability tokens** (`OP_CONTRACT_READ` / `OP_CONTRACT_SIGN`), **k-of-n `ContractStateTip` signing**, and new shared **`CONTRACT_MESSAGE` body types** (group journal catch-up, capability grants, withdrawal request/witness).
> 
> **Peer / wire:** `contract:message` events now carry **`wireMessage` / `messageHex`** for journal attach; contract publish/patch allow-lists are tightened (wire signer must match body authority arrays; publisher no longer auto-granted). Document private relay gains **`_sendPrivateRelayedDocumentRequest`**; peeled/relayed document requests stay observe-only via delivery options. **`SIDECHAIN_STATE_PATCH`** gains optional **`patchesCanonical`** for multi-op RFC6902 fidelity; message **`decodeBody`** defaults missing optional fields.
> 
> **Chores:** root **`.npmrc` `allow-git=all`** and install reporting for npm 12+ GitHub consumers; developer docs note npm 12+ requirement.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a97e7523a28be8b2f61311db1ea25dc6aec4dd7f. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: functions/contractTaproot.js [hunk 1]
```diff
@@ -0,0 +1,851 @@
...
+  if (!ms.length) throw new Error('leafScript is required.');
+
+  const dest = String(destinationAddress || '').trim();
+  if (!dest) throw new Error('destinationAddress is required.');
+
+  const afterLock = normalizeLock(opts.after);
+  let sequence = opts.sequence != null ? Number(opts.sequence) : undefined;
+  let locktime;
+  if (afterLock && afterLock.type === 'csv') {
+    if (sequence == null) sequence = afterLock.value;
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractTaproot.js:1047]: Dismissing: BIP68 `SEQUENCE_LOCKTIME_TYPE_FLAG` (`0x00400000`) selects *512-second* units. Block-based CSV uses the raw block count in bits 0–15 with bit 31 clear — which is what `prepareLeafPsbt` already encodes. Covered by `prepareLeafPsbt encodes BIP68 block-based CSV in low 16 bits (no TYPE_FLAG)`.

File: functions/documentRegistrySidechain.js [hunk 2]
```diff
@@ -120,15 +124,6 @@ function applyRegistryUpdateFields (state, fields, policy = null) {
   if (!fields || typeof fields.catalogCanonical !== 'string' || !fields.catalogCanonical) {
     return { ok: false, error: 'catalogCanonical required' };
   }
-  let catalog;
-  try {
-    catalog = JSON.parse(fields.catalogCanonical);
-  } catch (err) {
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/documentRegistrySidechain.js:126]: Already fixed in tree: `applyRegistryUpdateFields` prefers a non-empty `patchesCanonical` RFC6902 array *before* requiring `catalogCanonical` (see lines ~185–207). Multi-op patches with an empty catalog placeholder are accepted.

File: tests/contractTaproot.unit.js [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    assert.strictEqual(timelockMature({ type: 'csv', blocks: 10 }, { utxoAgeBlocks: 10 }), true);
+    assert.strictEqual(timelockMature({ type: 'cltv', height: 100 }, { tipHeight: 99 }), false);
+    assert.strictEqual(timelockMature({ type: 'cltv', height: 100 }, { tipHeight: 100 }), true);
+  });
+
+  it('normalizeLock rejects CSV above 65535', function () {
+    assert.ok(normalizeLock({ type: 'csv', blocks: 65535 }));
+    assert.strictEqual(normalizeLock({ type: 'csv', blocks: 65536 }), null);
+    assert.strictEqual(normalizeLock({ type: 'csv', blocks: 0 }), null);
+  });
```
[the remark this answers is not in the document]
Comment (martindale) [reply tests/contractTaproot.unit.js:380]: Already fixed: `normalizeLock` enforces BIP65 height `< 500000000` and unix `≥ 500000000`, with unit tests for the boundary values.

File: functions/contractProgramBind.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  const programHash = String(program.programHash || program.hash && program.hash() || '')
+    .trim()
+    .toLowerCase();
+  if (!/^[0-9a-f]{64}$/.test(programHash)) {
+    return { ok: false, error: 'program.programHash unavailable' };
+  }
+  const expected = String(program.runCommitmentHex({
+    tip: run.tip,
+    stack: run.stack || run.trace || []
+  })).toLowerCase();
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractProgramBind.js:186]: Fixed earlier + regression added: `assertMachineRunMatches` prefers `run.trace`; new test seeds the Machine stack before `runProgram` and asserts match still succeeds.

File: functions/contractTaproot.js [hunk 5]
```diff
@@ -0,0 +1,1525 @@
...
+  const softMode = String(opts.softMode || 'publisher').toLowerCase();
+  let softTier;
+  if (softMode === 'reduced' || softMode === 'threshold') {
+    const softThr = Math.max(
+      1,
+      Number(opts.softThreshold) || Math.max(1, Math.ceil(threshold / 2))
+    );
+    softTier = {
+      id: 't1-soft',
+      threshold: Math.min(softThr, sorted.length),
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractTaproot.js]: Fixed: `synthesizeDefaultLadder` now throws when `threshold` / `softThreshold` exceeds the unique key count after dedupe (no silent `Math.min` clamp). Unit coverage extended.

File: types/contract.js [hunk 6]
```diff
@@ -320,16 +320,155 @@ class Contract extends Service {
...
+      if (!this._state) this._state = { content: {} };
+      this._state.content = next.content;
+      if (this.settings && this.settings.state) {
+        this.settings.state = next.content;
+      }
+      observation.applied = true;
+      observation.balances = next.content.balances;
+    } else if (observation.ok) {
+      observation.applied = false;
+    }
```
[the remark this answers is not in the document]
Comment (martindale) [reply types/contract.js:390]: Fixed: `_handleBitcoinTransaction` merges payment observation into the existing `_state.content` object in place so `monitor.observe` stays attached.

File: types/machine.js [hunk 7]
```diff
@@ -271,11 +352,39 @@ class Machine extends Actor {
...
+            ? result.tip
+            : (result.stack && result.stack.length ? result.stack[result.stack.length - 1] : null),
+          trace: result.trace || [],
+          runCommitmentHex: result.runCommitmentHex || null,
+          programHash: result.programHash || null,
+          stepsExecuted: result.stepsExecuted,
+          error: result.error,
+          program: result.program || prog
+        };
+      }
```
[the remark this answers is not in the document]
Comment (martindale) [reply types/machine.js:389]: Fixed: `runExecutionProgram` / `Machine.runProgram` now apply genesis `primitives.opcodes` and enforce the allow-list on `FabricOpcode` steps. Added unit coverage.

File: functions/contractMessageAccumulate.js [hunk 8]
```diff
@@ -0,0 +1,1065 @@
...
+    }
+    const token = meta.capabilityToken || meta.token;
+    if (!allowed.has(authorXOnly) && !authorHasReadToken(doc, authorXOnly, token)) {
+      return { ok: false, error: 'author not in contract reader/signer set' };
+    }
+    return { ok: true };
+  }
+
+  const needsSigner = SIGNER_MUTATION_TYPES.includes(type);
+  if (!needsSigner) return { ok: true };
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractMessageAccumulate.js]: Fixed: `SIGNER_MUTATION_TYPES` now includes `GroupChangeProposal`, `GroupChangeVote`, `FederationContractInvite`(+Response), and `GroupJournalRequest`, so outsiders cannot advance tip clock / `stateDigest` with AMP-signed spam. Regression test added.

File: functions/contractMessageAccumulate.js [hunk 9]
```diff
@@ -0,0 +1,1065 @@
...
+    members: memberList,
+    signers: signerList,
+    messages,
+    withdrawals: Object.keys(withdrawals).sort().map((k) => withdrawals[k])
+  };
+  // Delivery ACKs are persisted in entries but do not advance tip clock / digest.
+  const tipEntries = sorted.filter((row) => !isDeliveryAckType(row.type));
+  const state = {
+    version: 1,
+    clock: tipEntries.length,
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractMessageAccumulate.js:894]: Addressed via the authorizeIngest fail-closed change for governance/invite/journal-request types (same fix as the coupled tip-clock thread).

File: functions/contractMessageAccumulate.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractMessageAccumulate.js]: Duplicate of the authorizeIngest tip-integrity finding — fixed by expanding `SIGNER_MUTATION_TYPES` + regression test.

File: functions/contractMessageAccumulate.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractMessageAccumulate.js:894]: Duplicate tip-clock thread — resolved by the same authorizeIngest change.

File: functions/blindedExecutionCircuit.js [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+  const coordinationRoot = session.coordinationRoot || computeCoordinationRoot(session.decisions);
+  const circuitCommitment = digestCanonical({
+    version: SESSION_VERSION,
+    kind: 'BlindedExecutionFinal',
+    baseCircuitCommitment: session.circuitCommitment,
+    coordinationRoot,
+    runCommitmentHex: runCommitmentHex || null
+  });
+
+  const hashlockPreimage = Buffer.from(circuitCommitment, 'hex');
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/blindedExecutionCircuit.js:472]: Fixed: hashlock leaf now requires `pubkeyHex` (publisher) so witness shape is `preimage+sig`. Public circuit commitment alone can no longer sweep `bound.address`; spend needs the authorized Schnorr sig + preimage. Tests updated accordingly.

File: functions/blindedExecutionCircuit.js [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+  if (!proposalMerkleRoot || !/^[0-9a-f]{64}$/.test(proposalMerkleRoot)) {
+    throw new Error('recordProposalDecision: proposal merkleRoot missing');
+  }
+
+  if (proposalPayload.contractId != null && session.circuitId &&
+      String(proposalPayload.contractId) !== String(session.circuitId)) {
+    throw new Error('recordProposalDecision: proposal contractId does not match session');
+  }
+
+  const entry = {
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/blindedExecutionCircuit.js:400]: Fixed: `recordProposalDecision` rejects actors outside garbler ∪ evaluators; `finalizeBlindedExecution` now requires ≥1 **evaluator** accept (garbler-only accept is insufficient).

File: functions/contractMessageAccumulate.js [hunk 12]
```diff
@@ -0,0 +1,1096 @@
...
+      duplicate: false,
+      entry: null,
+      tip: tipFromDoc(loadDoc(store, id)),
+      error: e.message || 'invalid buffer'
+    };
+  }
+
+  let msg;
+  try {
+    msg = Message.fromBuffer(buffer);
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/contractMessageAccumulate.js:1095]: Fixed: `ingestMessageBuffer` rejects buffers larger than `HEADER_SIZE + MAX_MESSAGE_SIZE` before `Message.fromBuffer` / verify / JSON — matches the Peer wire gate. Added a unit test.

File: constants.js [hunk 13]
```diff
@@ -16,14 +16,60 @@ const FABRIC_USER_AGENT = 'Fabric Core 0.1.0 (@fabric/core#v0.1.0-RC1)';
...
+/**
+ * Fabric identity path: m/44'/7778'/account'/0/index
+ * @param {number} [account=0]
+ * @param {number} [index=0]
+ * @returns {string}
+ */
+function fabricIdentityDerivationPath (account = 0, index = 0) {
+  const a = Math.max(0, Math.floor(Number(account) || 0));
+  const i = Math.max(0, Math.floor(Number(index) || 0));
+  return `m/44'/${FABRIC_COIN_TYPE}'/${a}'/0/${i}`;
```
[the remark this answers is not in the document]
Comment (martindale) [reply constants.js]: Fixed: `assertBip32ChildIndex` rejects non-finite / fractional / out-of-domain values; used by Fabric/Bitcoin path helpers and `Identity` constructor / `loadAccountByID`.

File: functions/blindedExecutionCircuit.js [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+    decision,
+    actorPubkey,
+    proposalMerkleRoot,
+    at: opts.at != null ? Number(opts.at) : Date.now()
+  };
+
+  const decisions = session.decisions.concat([entry]);
+  const coordinationRoot = computeCoordinationRoot(decisions);
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/blindedExecutionCircuit.js:409]: Fixed (same as the unauthenticated-accept High): session role membership is enforced in `recordProposalDecision`, with a test for outsider rejection.

File: functions/executionProgramRunner.js [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+    // Prefer explicit opts / genesis list when an existing Machine is injected.
+    const existing = machine.allowedOpcodes || [];
+    if (!existing.length) {
+      machine.setAllowedOpcodes(allowList);
+    } else {
+      allowList = existing;
+    }
+  } else if (!allowList.length && options.machine instanceof Machine) {
+    allowList = machine.allowedOpcodes || [];
+  }
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/executionProgramRunner.js:196]: Fixed: when an explicit opts/genesis allow-list is provided, `runExecutionProgram` now prefers it over any preloaded `Machine.allowedOpcodes` (matches the in-code comment).

File: functions/fabricListenInterface.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+  const keys = Array.isArray(opts.envKeys) && opts.envKeys.length
+    ? opts.envKeys
+    : ['FABRIC_INTERFACE', 'FABRIC_PEER_INTERFACE'];
+  for (const key of keys) {
+    const v = String(env[key] || '').trim();
+    if (v) return v;
+  }
+  const explicit = String(opts.interface || opts.host || '').trim();
+  if (explicit) return explicit;
+  return opts.fallback != null ? String(opts.fallback) : '0.0.0.0';
```
[the remark this answers is not in the document]
Comment (martindale) [reply functions/fabricListenInterface.js]: JSDoc aligned with the tested precedence (env → settings → fallback). Changing to settings-first would break `tests/fabric.listenInterface.js` and host NIC pinning via `FABRIC_INTERFACE`.

File: IDENTITY.md [hunk 17]
```diff
@@ -17,10 +21,11 @@ Identity #2 [1]: m/44'/7778'/1'/0/0
 ...
 ```
 
-Fabric will encode each transaction using the address index field, as specified in BIP 44:
+Fabric will encode each identity address index as specified by BIP 44 for the Fabric coin type
+(not Bitcoin funds):
 
 ```
-Transaction #1 [1]: m/44'/7778'/0'/1/0
+Identity account 0, address 1: m/44'/7778'/0'/0/1
```
[the remark this answers is not in the document]
Comment (martindale) [reply IDENTITY.md]: Fixed: added `text` language tags on the IDENTITY.md fenced blocks (MD040).

File: package.json [hunk 18]
```diff
@@ -53,7 +53,8 @@
     "docs/**/*.md"
   ],
   "engines": {
-    "node": "24.15.0"
+    "node": "24.15.0",
+    "npm": ">=12.0.0"
```
[the remark this answers is not in the document]
Comment (martindale) [reply package.json:57]: Intentional: `engines.npm: >=12` is required for git-dep install behavior; DEVELOPERS.md § setup already documents upgrading past Node 24.15.0’s bundled npm 11.x. Leaving engines as-is.

File: tests/playnet.contract.publish.integration.js [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+
+    let publishes = 0;
+    peer.on('contract:publish', () => { publishes++; });
+
+    const wire = Message.fromVector(['CONTRACT_PUBLISH', JSON.stringify(definition)]).signWithKey(owner);
+    peer._handleGenericMessage({
+      type: 'CONTRACT_PUBLISH',
+      actor: { publicKey: ownerPub },
+      object: definition
+    }, { name: '127.0.0.1:18444' });
```
[the remark this answers is not in the document]
Comment (martindale) [reply tests/playnet.contract.publish.integration.js]: Fixed: offline playnet test now feeds the signed frame through `_handleFabricMessage`.

File: tests/playnet.contract.publish.integration.js [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+  (runLive ? it : it.skip)('dials configured peers and emits CONTRACT_PUBLISH', async function () {
+    this.timeout(60000);
+    const peers = String(process.env.FABRIC_PLAYNET_PEERS || '127.0.0.1:7777')
+      .split(',')
+      .map((s) => s.trim())
+      .filter(Boolean);
+    const mnemonic = String(process.env.FABRIC_MNEMONIC ||
+      'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about').trim();
+
```
[the remark this answers is not in the document]
Comment (martindale) [reply tests/playnet.contract.publish.integration.js]: Fixed: live playnet test skips unless `FABRIC_MNEMONIC` is explicitly set (no public fallback).

File: tests/playnet.contract.publish.integration.js [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+    const msg = Message.fromVector(['CONTRACT_PUBLISH', JSON.stringify(definition)]).signWithKey(peer.key);
+    peer.broadcast(msg.toBuffer());
+
+    // Brief observe, then re-publish once.
+    await new Promise((r) => setTimeout(r, 1000));
+    const msg2 = Message.fromVector(['CONTRACT_PUBLISH', JSON.stringify(definition)]).signWithKey(peer.key);
+    peer.broadcast(msg2.toBuffer());
+    await new Promise((r) => setTimeout(r, 1000));
+
+    await peer.stop();
```
[the remark this answers is not in the document]
Comment (martindale) [reply tests/playnet.contract.publish.integration.js]: Fixed: live playnet peer `stop()` runs in a `finally` block.
